### PR TITLE
join: remove crash! macro

### PR DESCRIPTION
**Pull request by cswn**
This is related to issue https://github.com/uutils/coreutils/issues/5487. Removes the `crash!` macro from join.

### CODE REVIEW

File: src/uu/join/src/join.rs [hunk 1]
```diff
@@ -334,28 +334,34 @@ impl<'a> State<'a> {
...
-        let f = if name == "-" {
+    ) -> Result<State<'a>, JoinError> {
+        let file_buf = if name == "-" {
             Box::new(stdin.lock()) as Box<dyn BufRead>
         } else {
             match File::open(name) {
                 Ok(file) => Box::new(BufReader::new(file)) as Box<dyn BufRead>,
-                Err(err) => crash!(1, "{}: {}", name.maybe_quote(), err),
+                Err(err) => {
+                    return Err(JoinError::UnorderedInput(format!(
```
Comment (tertsdiepraam) [inline src/uu/join/src/join.rs:344]: `UnorderedInput` is probably not quite right is it? It is the most convenient solution and it works, but it doesn't make sense :smile:

`UError` has a way of adding some context to an io error like so:

```rs
let file = File::open(name).map_err_context(|| name.maybe_quote())?;
```

Then, the function should return a `UResult` (and all the callers of this function as well). I think that should be fairly easy to do.

That is how other utils usually do this. Could you give that a try?

**Review by tertsdiepraam — COMMENTED** [review]
Nice! I think you're on the right track and the solution is correct, but I do have a suggestion to make it a bit more like the other utils (and not having to use the `UnorderedInput` variant).

File: src/uu/join/src/join.rs (hunk 1 above)
Comment (cswn) [reply src/uu/join/src/join.rs:344]: Absolutely, thanks for the feedback!

**Review by tertsdiepraam — COMMENTED** [review]
Excellent work! This is exactly what I had in mind. Because we now use `UResult` I think we can also remove this bit:

https://github.com/uutils/coreutils/blob/a064c886566f810c3e58b1b1153762e772e35567/src/uu/join/src/join.rs#L704-L707C6

and just make that a call to `exec` and return the result. And then we can merge this.

**Review by tertsdiepraam — APPROVED** [review]
Yes, that's it! I'll wait for the CI but seems ready to be merged. Thanks!

**Comment by cakebaker** [comment]
Thanks!

**Pull request merged**